go/vt/sqlparser/expression_rewriting.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2019 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sqlparser
+
+import (
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+// PrepareAST will normalize the query
+func PrepareAST(in Statement, bindVars map[string]*querypb.BindVariable, prefix string) (*RewriteASTResult, error) {
+	Normalize(in, bindVars, prefix)
+	return RewriteAST(in)
+}
+
+// RewriteAST rewrites the whole AST, replacing function calls and adding column aliases to queries
+func RewriteAST(in Statement) (*RewriteASTResult, error) {
+	er := new(expressionRewriter)
+	Rewrite(in, er.goingDown, er.comingUp)
+
+	return &RewriteASTResult{
+		AST:              in,
+		NeedLastInsertID: er.lastInsertID,
+		NeedDatabase:     er.database,
+	}, nil
+}
+
+// RewriteASTResult contains the rewritten ast and meta information about it
+type RewriteASTResult struct {
+	AST              Statement
+	NeedLastInsertID bool
+	NeedDatabase     bool
+}
+
+type expressionRewriter struct {
+	lastInsertID, database bool
+	err                    error
+	aliases                []*AliasedExpr
```

Comment:
I now see why you were trying to add aliases for every expression :).

If you were already willing to pay the price of generating an alias for every select expression, there is a simpler way: As soon as you encounter an unaliased *AliasedExpression, generate an alias for it locally. Then create a new `RewriteASTResult` and invole a `Rewrite` on it. If it returns saying that something has changed, then you substitute the new alias after it returns. And always return `false`.

This avoids the need to maintain a separate `aliases` list, and also there is no need for a `comingUp` function.

Reply:
Nice idea. Much cleaner.